File: cpp/cpp_100-999/209.cc
// C++
// leetcode 209
// https://leetcode-cn.com/problems/minimum-size-subarray-sum/

class Solution {
public:
    int minSubArrayLen(int target, vector<int>& nums) {
    	int left = 0, right = 0, sum = nums[0], n = nums.size();
   		int res = INT_MAX;
    	while(right < n) {
    		while(target > sum) {
    			right++;
    			if(right < n) {
    				sum += nums[right];
    			} else {
    				break;
    			}
    		}
    		if(target <= sum) {
    			res = min(res, right - left + 1);
    			sum -= nums[left];
                left++;
    		}
    	}
    	return res == INT_MAX ? 0 : res;
    }
};

File: Go/Go_100-999/209.go
// Go
// leetcode 209
// https://leetcode-cn.com/problems/minimum-size-subarray-sum/

func minSubArrayLen(target int, nums []int) int {
	res := 1000_000
	n := len(nums)
	curSum :=0
	for l, r:=0, 0; l < n; {
		for r < n && curSum < target {
			curSum += nums[r]
			r++
		}
		
		if curSum < target {
			break
		}
		
        for l < n && curSum >= target {
            curSum -= nums[l]
            l++
        }
        res = min(res, r-l+1)
		
	}
	
	if res == 1000_000 {
		return 0
	}
	return res
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}